core/embed/prodtest/main.c

```diff
@@ -513,9 +603,443 @@ static void test_otp_write_device_variant(const char *args) {
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  if (!set_metadata(OID_CERT_FIDO, &metadata)) {
+    return;
+  }
+
+  // Set metadata for device private key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = ACCESS_PAIRED;
+  metadata.key_usage = KEY_USE_SIGN;
+  if (!set_metadata(OID_KEY_DEV, &metadata)) {
+    return;
+  }
+
+  // Set metadata for FIDO attestation private key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = ACCESS_PAIRED;
+  metadata.key_usage = KEY_USE_SIGN;
+  if (!set_metadata(OID_KEY_FIDO, &metadata)) {
+    return;
+  }
+
+  // Set metadata for pairing key.
+  memzero(&metadata, sizeof(metadata));
+  metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  metadata.change = OPTIGA_ACCESS_NEVER;
+  metadata.read = OPTIGA_ACCESS_NEVER;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  metadata.data_type = TYPE_PTFBIND;
+  if (!set_metadata(OID_KEY_PAIRING, &metadata)) {
+    return;
+  }
+
+  vcp_println("OK");
+}
+
+static secbool is_optiga_locked(void) {
+  const uint16_t oids[] = {OID_CERT_DEV, OID_CERT_FIDO, OID_KEY_DEV,
+                           OID_KEY_FIDO, OID_KEY_PAIRING};
+
+  optiga_metadata locked_metadata = {0};
+  locked_metadata.lcso = OPTIGA_LCS_OPERATIONAL;
+  for (size_t i = 0; i < sizeof(oids) / sizeof(oids[0]); ++i) {
+    uint8_t metadata_buffer[258] = {0};
+    size_t metadata_size = 0;
+    optiga_result ret =
+        optiga_get_data_object(oids[i], true, metadata_buffer,
+                               sizeof(metadata_buffer), &metadata_size);
+    if (OPTIGA_SUCCESS != ret) {
+      vcp_println("ERROR optiga_get_metadata error %d for OID 0x%04x.", ret,
+                  oids[i]);
+      return secfalse;
+    }
+
+    optiga_metadata stored_metadata = {0};
+    ret =
+        optiga_parse_metadata(metadata_buffer, metadata_size, &stored_metadata);
+    if (OPTIGA_SUCCESS != ret) {
+      vcp_println("ERROR optiga_parse_metadata error %d.", ret);
+      return secfalse;
+    }
+
+    if (!optiga_compare_metadata(&locked_metadata, &stored_metadata)) {
+      return secfalse;
+    }
+  }
+
+  return sectrue;
+}
+
+static void optigaid_read(void) {
+  uint8_t optiga_id[27] = {0};
+  size_t optiga_id_size = 0;
+
+  optiga_result ret = optiga_get_data_object(
+      OID_OPTIGA_UID, false, optiga_id, sizeof(optiga_id), &optiga_id_size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_get_data_object error %d for 0x%04x.", ret,
+                OID_OPTIGA_UID);
+    return;
+  }
+
+  vcp_print("OK ");
+  vcp_println_hex(optiga_id, optiga_id_size);
+}
+
+static void cert_read(uint16_t oid) {
+  uint8_t cert[1024] = {0};
```

Comment:
I don't know where this number came from, but it seems to me that the certificate can have up to 1727 bytes. See [this](https://github.com/Infineon/optiga-trust-m/blob/develop/documents/OPTIGA%E2%84%A2%20Trust%20M%20Solution%20Reference%20Manual.md#link24b48059_db81_40f5_8b65_7afca4918ab1).

Reply:
agreed, let's bump this to 2kB to be on the safe side.

(do we have stack big enough for this? maybe the big buffers should be marked `static` to go live somewhere other than the stack? or does C take care of this automagically?)

Reply:
Done in f92db56fc8319aa573536e39396597e0e99ebd46. The zero-initialization is not needed, it's just a general rule to initialize stack variables with zeros to ensure that in case some sensitive data is left in the freed stack, it gets erased.